Auto-assign egress IPs from EgressCIDRs

danwinship opened the pull request:
Building on #20155 (the first commit), which moves the EgressIP-tracking code from node/ to common/, and https://github.com/openshift/api/pull/64 (the second commit), which adds HostSubnets.EgressCIDRs, this adds support to the common egressip code for tracking changes to EgressCIDRs and figuring out acceptable allocations of EgressIPs based on that; and adds support to master for actually updating HostSubnet objects to reflect these new figured-out allocations.

(This PR does not include HA.)

knobunc commented:
@deads2k we will need approval for the api glide change, once the api change goes in.

danwinship commented:
/test integration

danwinship commented:
/test extended_conformance_install

danwinship commented:
/test integration

pravisankar commented:
/retest

danwinship commented:
/retest

danwinship commented:
/retest

dcbw commented:
/test integration
/test cmd

squeed commented on pkg/network/master/egressip.go:
```diff
@@ -0,0 +1,95 @@
+		eim.updatedAgain = true
+	} else {
+		eim.updatePending = true
+		go utilwait.PollInfinite(time.Second, eim.maybeDoUpdateEgressCIDRs)
```
Will you have a deadlock here? eim.Lock is already held, if I understand this correctly

danwinship replied:
between UpdateEgressCIDRs() and maybeDoUpdateEgressCIDRs()? No, because the latter is called in a goroutine, after the former returns

dcbw replied:
@danwinship so basically, if UpdateEgressCIDRs() gets called multiple times in the same second, the next maybeDoUpdateEgressCIDRs() will be skipped and it'll wait until it gets at least one second of no activity before it actually does the stuff in maybeDoUpdateEgressCIDRs(), right?

squeed replied:
ah yes, makes sense. assuming I understand this correctly, nothing should be skipped because it will just wait for the lock.

danwinship commented:
/test integration

squeed commented on pkg/network/master/egressip.go:
```diff
@@ -0,0 +1,95 @@
+	}
+}
+
+func (eim *egressIPManager) maybeDoUpdateEgressCIDRs() (bool, error) {
```
Could you add a quick docblock explaining what this does?

squeed commented on pkg/network/master/egressip.go:
```diff
@@ -0,0 +1,95 @@
+	return true, nil
+}
+
+func (eim *egressIPManager) ClaimEgressIP(vnid uint32, egressIP, nodeIP string) {
```
why are these functions no-ops? If there's a good reason, could you add a comment?

danwinship replied:
It's part of the `common.EgressIPWatcher` interface. I was trying to share code between the node and master, and ended up doing it in a bad way. I'm going to refactor this later. For now though, the master needs to provide no-op implementations of these methods, and the node needs to provide a no-op implementation of `UpdateEgressIPs`.

squeed commented:
A few questions, and a desire for more comments, but LGTM otherwise.

danwinship commented:
integration is failing due to mismatch between openshift/api and openshift/origin master

pravisankar commented on pkg/network/common/egressip.go:
```diff
@@ -177,10 +187,11 @@ func (eit *EgressIPTracker) UpdateHostSubnetEgress(hs *networkapi.HostSubnet) {
 
 	node := eit.nodesByNodeIP[hs.HostIP]
```
Do we have an issue in this case?
- Got add HostSubnet event, h1 with host IP ip1 and egress IP e1
- So we populated nodesByNodeIP[ip1] = { nodeIP: ip1, requestedIPs: [ e1 ], ... }
- Got modified HostSubnet event for h1 and host IP changed from ip1 to ip2
- We populate nodesByNodeIP[ip2]  = { nodeIP: ip2, requestedIPs: [ e1 ] ...}
   So now egress IP e1 is used by ip1 and ip2 according to the map even though ip1 is no longer valid.
   We need to delete obsoleted nodesbyNodeIP[ip1] and we could do that by keeping track of node UID --> node IP map?

danwinship replied:
Doh. Yes, that applies to the existing "semi-automatic" code too though, so that's separate from this new feature. I'll do another PR to fix that after this.

pravisankar commented on pkg/network/common/egressip.go:
```diff
@@ -301,6 +329,13 @@ func (eit *EgressIPTracker) syncEgressIPs() {
+		if eit.nodesWithCIDRs > 0 {
+			eit.watcher.UpdateEgressCIDRs()
+		}
+	}
```
Don't this need to be:
```
    if eit.updateEgressCIDRs {
        eit.updateEgressCIDRs = false
        eit.watcher.UpdateEgressCIDRs()
    }
```
Use case:
- We got a node with egress CIDR, so nodesWithCIDRs = 1 and watcher.UpdateEgressCIDRs will update corresponding hs.EgressIPs
- Now egress CIDR got removed from the same node, so nodeWithCIDRs = 0 and in this case we do need to call watcher.UpdateEgressCIDRs to remove old hs.EgressIPs?

danwinship replied:
I debated about this. The problem is that `EgressCIDRs == nil` is defined to mean "don't automatically set EgressIPs on this node", so that behavior would be inconsistent with the normal behavior. (And it would cause problems if the controller restarted while the apiserver was still running, in which case the non-nil-to-nil transition might happen while the controller wasn't running, and then afterward it would have no idea that it was supposed to remove those EgressIPs.) And anyway, it's easy enough for the user to just unset EgressIPs themselves when they unset EgressCIDRs if that's the behavior they wanted.

I agree it's unexpected though. Maybe there should have been another field, `DoAutomaticEgressIPAssignment bool`, rather than overloading that into `EgressCIDRS == nil` / `!= nil`...

danwinship commented on pkg/network/apis/network/validation/validation.go:
```diff
@@ -174,6 +174,12 @@ func ValidateHostSubnet(hs *networkapi.HostSubnet) field.ErrorList {
+
 	return allErrs
 }
 
```
No, because that would prevent the controller from being able to update EgressIPs too

pravisankar commented on pkg/network/common/egressip.go:
```diff
@@ -402,3 +441,100 @@ func (eit *EgressIPTracker) Ping(ip string, timeout time.Duration) bool {
+				if bestNode != "" {
+					otherNodes = true
+					if len(allocation[bestNode]) < len(egressIPs) {
+						continue
```
break instead of continue?
Continue may not help because even if other cidr on the node matched the ip, len(egressIPs) for the node won't change and it will be greater than allocation[bestNode].

knobunc commented:
/retest

danwinship commented:
/retest

danwinship commented:
@dcbw 
> so basically, if UpdateEgressCIDRs() gets called multiple times in the same second, the next maybeDoUpdateEgressCIDRs() will be skipped...

yes; I added some comments explaining that after Casey's review but I think you reviewed an older commit.

Made the code reorg you suggested and rebased to master

danwinship commented:
/test gcp

squeed commented:
/lgtm

danwinship commented:
/retest

deads2k commented:
bump looks clean

/approve

danwinship commented:
/retest

liggitt commented:
thanks for getting this updated. ready for tagging for merge?

danwinship commented:
@liggitt yes

liggitt commented:
readding lgtm from https://github.com/openshift/origin/pull/20258#issuecomment-404876248

liggitt commented:
/retest

danwinship commented:
/retest

danwinship commented:
/test gcp

knobunc commented:
/test gcp
